### PR TITLE
fix: resources not freed immediately after stage finished

### PR DESCRIPTION
**Pull request by zhujian7**
**What this PR does / why we need it**:

- delete pod immediately after stage finished
- create pod retry on exceeded quota

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @cd1989 @supereagle 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/workflow/controller/config.go [hunk 1]
```diff
@@ -48,6 +48,9 @@ type WorkflowControllerConfig struct {
 	NotificationURL string `json:"notification_url"`
 	// DindSettings is settings for Docker in Docker
 	DindSettings DindSettings `json:"dind"`
+
+	// RetrySettins is settings for retry mechanism.
```
Comment (supereagle) [inline pkg/workflow/controller/config.go]: Typo of `RetrySettins`

File: pkg/workflow/controller/config.go [hunk 2]
```diff
@@ -48,6 +48,9 @@ type WorkflowControllerConfig struct {
...
+	RetrySettings RetrySettings `json:"retry"`
```
Comment (supereagle) [inline pkg/workflow/controller/config.go]: I'd prefer `Retry`

File: pkg/workflow/controller/config.go [hunk 3]
```diff
@@ -92,6 +95,15 @@ type DindSettings struct {
 	InsecureRegistries []string `json:"insecure_registries"`
 }
 
+// RetrySettings is settings for controller retry mechanism.
+// For example, count: 6, period: 10 defines try to do something every 10 seconds and do it 6 times.
+type RetrySettings struct {
+	// Count is the max retry times.
+	Count int `json:"count"`
+	// Period is the retry mechanism interval time, unit seconds.
+	Period int `json:"period"`
```
Comment (supereagle) [inline pkg/workflow/controller/config.go]: Maybe `interval` is better

File: pkg/workflow/controller/config.go [hunk 4]
```diff
@@ -92,6 +95,15 @@ type DindSettings struct {
 	InsecureRegistries []string `json:"insecure_registries"`
 }
 
+// RetrySettings is settings for controller retry mechanism.
+// For example, count: 6, period: 10 defines try to do something every 10 seconds and do it 6 times.
+type RetrySettings struct {
```
Comment (supereagle) [inline pkg/workflow/controller/config.go]: I'd prefer `Retry`

File: pkg/workflow/workflowrun/workload.go [hunk 5]
```diff
@@ -75,8 +78,21 @@ func (p *WorkloadProcessor) processPod() error {
...
-	// Create the generated pod.
-	po, err = p.clusterClient.CoreV1().Pods(pod.GetExecutionContext(p.wfr).Namespace).Create(po)
+	// Create the generated pod with retry on exceeded quota.
+	// Here is a litter tricky. Cyclone will delete stage related pod to release cpu/memory resource when stage have
+	// been finished, but pod deletion needs some time, so retry on exceeded quota gives the time to waiting previous
+	// stage pod deletion.
+	backoff := wait.Backoff{
+		Steps:    controller.Config.RetrySettings.Count,
+		Duration: time.Duration(controller.Config.RetrySettings.Period) * time.Second,
+		Factor:   1.0,
```
Comment (supereagle) [inline pkg/workflow/workflowrun/workload.go]: If the resources are not sufficient, Cyclone should wait more time before next retry. I think this should be added into retry config, and set a reasonable default value.

File: pkg/util/retry/retry.go [hunk 6]
```diff
@@ -22,3 +25,27 @@ func OnError(backoff wait.Backoff, fn func() error) error {
 	}
 	return err
 }
+
+// OnExceededQuota executes the provided function repeatedly, retrying if the server returns an exceeded quota error.
+func OnExceededQuota(backoff wait.Backoff, fn func() error) error {
+	var lastConflictErr error
```
Comment (cd1989) [inline pkg/util/retry/retry.go]: variable name ?

**Comment by zhujian7** [comment]
Discussed offline: WorkflowRun retry mechanism needs a deeper design, here remove the retry configuration to simplify usage.

**Comment by zhujian7** [comment]
@supereagle PTAL

File: pkg/workflow/workflowrun/workload.go [hunk 7]
```diff
@@ -75,8 +77,21 @@ func (p *WorkloadProcessor) processPod() error {
...
+	// Here is a litter tricky. Cyclone will delete stage related pod to release cpu/memory resource when stage have
+	// been finished, but pod deletion needs some time, so retry on exceeded quota gives the time to waiting previous
+	// stage pod deletion.
+	backoff := wait.Backoff{
+		Steps:    3,
+		Duration: 5 * time.Second,
+		Factor:   1.5,
+		Jitter:   0.1,
+	}
+	origin := *po
```
Comment (cd1989) [inline pkg/workflow/workflowrun/workload.go]: deep copy or directly pointer assign

**Comment by zhujian7** [comment]
@cd1989 PTAL

**Pull request merged**